### ISSUE TITLE
no-duplicate-case false positives with CallExpressions

### ISSUE THREAD

**Issue by matz3** [eslint/eslint#2499]
The `getHash` function of the `no-duplicate-case` rule does not include CallExpressions so the rule will report false positives when using function calls in case statements.

Some examples:

```
var a = 1,
    f = function(b) {
        return b ? {
            p1: 1
        } : {
            p1: 2
        };
    };

switch (a) {
    case f(true).p1:
        break;
    case f(true, false).p1:
        break;
    default:
        break;
}
```

```
var a = 1,
    f1 = function() {
        return {
            p1: 1
        }
    },
    f2 = function() {
        return {
            p1: 2
        }
    };

switch (a) {
    case f1().p1:
        break;
    case f2().p1:
        break;
    default:
        break;
}
```

```
var a = 1,
    f = function(s) {
        return {
            p1: s
        }
    };

switch (a) {
    case f(a + 1).p1:
        break;
    case f(a + 2).p1:
        break;
    default:
        break;
}
```
Related issue: https://github.com/eslint/eslint/issues/2038
